src/overlay/Peer.cpp

```diff
@@ -630,38 +710,256 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
+    {
+        assert(mCapacity.mTotalCapacity > 0);
+        if (mCapacity.mTotalCapacity > 0)
+        {
+            mCapacity.mTotalCapacity--;
+        }
+
+        if (mApp.getOverlayManager().isFloodMessage(stellarMsg))
+        {
+            if (mCapacity.mFloodCapacity == 0)
             {
-                try
-                {
-                    self->recvRawMessage(sm);
-                }
-                catch (CryptoError const& e)
-                {
-                    std::string err = fmt::format(
-                        FMT_STRING("Error RecvMessage T:{} cat:{} {} @{:d}"),
-                        mtype, cat, self->toString(), port);
-                    CLOG_ERROR(Overlay, "Dropping connection with {}: {}", err,
-                               e.what());
-                    self->drop("Bad crypto request",
-                               Peer::DropDirection::WE_DROPPED_REMOTE,
-                               Peer::DropMode::IGNORE_WRITE_QUEUE);
-                }
+                // TODO: for now, only test well-behaved peers, so throw for
+                // visibility. Later, drop this peer instead of crashing
+                throw std::runtime_error(
+                    fmt::format("{} sent unexpected flood message",
+                                mApp.getConfig().toShortString(getPeerID())));
             }
-            else
+
+            mCapacity.mFloodCapacity--;
+            if (mCapacity.mFloodCapacity == 0)
+            {
+                CLOG_TRACE(Overlay, "{}: no capacity for peer {}",
+                           mApp.getConfig().toShortString(
+                               mApp.getConfig().NODE_SEED.getPublicKey()),
+                           mApp.getConfig().toShortString(self->getPeerID()));
+            }
+        }
+    }
+
+    mApp.postOnMainThread(
+        [&app = mApp, weak, sm = StellarMessage(stellarMsg),
+         mtype = stellarMsg.type(), cat]() {
+            auto self = weak.lock();
+            if (!self)
             {
                 CLOG_TRACE(Overlay, "Error RecvMessage T:{} cat:{}", mtype,
                            cat);
+                return;
+            }
+
+            try
+            {
+                self->recvRawMessage(sm);
+            }
+            catch (CryptoError const& e)
+            {
+                std::string err = fmt::format(
+                    FMT_STRING("Error RecvMessage T:{} cat:{} {} @{:d}"), mtype,
+                    cat, self->toString(), app.getConfig().PEER_PORT);
+                CLOG_ERROR(Overlay, "Dropping connection with {}: {}", err,
+                           e.what());
+                self->drop("Bad crypto request",
+                           Peer::DropDirection::WE_DROPPED_REMOTE,
+                           Peer::DropMode::IGNORE_WRITE_QUEUE);
             }
+
+            // Notify on completion, schedule more reading if necessary
+            self->messageProcessed(sm);
```

Comment:
I think that if the message gets dropped by the scheduler, this doesn't get called so you end up never releasing capacity. What you were doing before (using shared ptr) was handling this case correctly

Reply:
good catch, this is fixed now with RAII